File: backend/crates/domain/coffret-format/src/control/object_name.rs
use std::fmt;

use coffret_model::{ContainerId, ControlObjectKind, Generation, ReplicaPosition};

use crate::error::{Error, Result};

/// The name a control object is stored under.
///
/// Control objects carry recognizable names because recovery discovers them by
/// name before any index exists:
///
/// ```text
/// jrn-<generation>.cfrt                                  Journal record
/// idx-<generation>.cfrt                                  Index Snapshot
/// key-<generation>-<set_digest>-r<index>-of-<count>.cfrt  Keyring replica
/// ```
///
/// A Journal record and an Index Snapshot are written once each, so their names
/// carry no replica position and they report [`ReplicaPosition::SINGLE`].
///
/// Numbers are spelled in decimal without leading zeros, so one object has
/// exactly one name: a reader that accepted `jrn-007.cfrt` as generation 7 would
/// let two names claim the same object.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ControlObjectName {
    /// A Journal record.
    Journal {
        /// Which generation of the Journal this record is.
        generation: Generation,
    },
    /// An Index Snapshot.
    IndexSnapshot {
        /// Which generation of the Index Snapshot this is.
        generation: Generation,
    },
    /// One replica of a Keyring.
    KeyringReplica {
        /// Which generation of the Keyring this replica belongs to.
        generation: Generation,
        /// The digest of the mapping the replica set carries.
        ///
        /// Its contents are the Keyring's business; a name only needs it to be
        /// a lowercase hex token, so that it cannot swallow the separators the
        /// rest of the name is parsed on.
        set_digest: String,
        /// Which replica this is, out of how many.
        replica: ReplicaPosition,
    },
}

/// The name prefix of a Journal record.
const JOURNAL_PREFIX: &str = "jrn-";
/// The name prefix of an Index Snapshot.
const INDEX_SNAPSHOT_PREFIX: &str = "idx-";
/// The name prefix of a Keyring replica.
const KEYRING_PREFIX: &str = "key-";

impl ControlObjectName {
    /// The name of one generation of the Journal.
    pub const fn journal(generation: Generation) -> Self {
        Self::Journal { generation }
    }

    /// The name of one generation of the Index Snapshot.
    pub const fn index_snapshot(generation: Generation) -> Self {
        Self::IndexSnapshot { generation }
    }

    /// The name of one replica of one generation of the Keyring.
    pub fn keyring_replica(
        generation: Generation,
        set_digest: &str,
        replica: ReplicaPosition,
    ) -> Result<Self> {
        // Lowercase only, as every hex spelling in coffret is: two spellings of
        // one digest would be two names for one object.
        let is_lowercase_hex = |byte: u8| matches!(byte, b'0'..=b'9' | b'a'..=b'f');
        if set_digest.is_empty() || !set_digest.bytes().all(is_lowercase_hex) {
            return Err(Error::MalformedObjectName {
                name: set_digest.to_owned(),
            });
        }
        Ok(Self::KeyringReplica {
            generation,
            set_digest: set_digest.to_owned(),
            replica,
        })
    }

    /// Which kind of control object this name belongs to.
    pub const fn kind(&self) -> ControlObjectKind {
        match self {
            Self::Journal { .. } => ControlObjectKind::Journal,
            Self::IndexSnapshot { .. } => ControlObjectKind::IndexSnapshot,
            Self::KeyringReplica { .. } => ControlObjectKind::Keyring,
        }
    }

    /// The generation the name encodes.
    pub const fn generation(&self) -> Generation {
        match self {
            Self::Journal { generation }
            | Self::IndexSnapshot { generation }
            | Self::KeyringReplica { generation, .. } => *generation,
        }
    }

    /// The replica position the name encodes.
    pub const fn replica(&self) -> ReplicaPosition {
        match self {
            Self::Journal { .. } | Self::IndexSnapshot { .. } => ReplicaPosition::SINGLE,
            Self::KeyringReplica { replica, .. } => *replica,
        }
    }

    /// The digest a Keyring replica's name carries, if this is one.
    pub fn set_digest(&self) -> Option<&str> {
        match self {
            Self::KeyringReplica { set_digest, .. } => Some(set_digest),
            _ => None,
        }
    }

    /// Reads a name back into the values it encodes.
    pub fn parse(name: &str) -> Result<Self> {
        let malformed = || Error::MalformedObjectName {
            name: name.to_owned(),
        };
        let body = name
            .strip_suffix(ContainerId::STORAGE_EXTENSION)
            .ok_or_else(malformed)?;

        if let Some(rest) = body.strip_prefix(JOURNAL_PREFIX) {
            return Ok(Self::journal(parse_generation(rest).ok_or_else(malformed)?));
        }
        if let Some(rest) = body.strip_prefix(INDEX_SNAPSHOT_PREFIX) {
            return Ok(Self::index_snapshot(
                parse_generation(rest).ok_or_else(malformed)?,
            ));
        }
        let rest = body.strip_prefix(KEYRING_PREFIX).ok_or_else(malformed)?;

        // The digest is hex, so it holds none of the `-` this splits on and the
        // five fields always land in the same places.
        let fields: Vec<&str> = rest.split('-').collect();
        let [generation, set_digest, index, of, count] = fields[..] else {
            return Err(malformed());
        };
        if of != "of" {
            return Err(malformed());
        }
        let generation = parse_generation(generation).ok_or_else(malformed)?;
        let index = index.strip_prefix('r').ok_or_else(malformed)?;
        let replica = ReplicaPosition::new(
            parse_u16(index).ok_or_else(malformed)?,
            parse_u16(count).ok_or_else(malformed)?,
        )?;
        Self::keyring_replica(generation, set_digest, replica).map_err(|_| malformed())
    }
}

impl fmt::Display for ControlObjectName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let extension = ContainerId::STORAGE_EXTENSION;
        match self {
            Self::Journal { generation } => {
                write!(f, "{JOURNAL_PREFIX}{generation}{extension}")
            }
            Self::IndexSnapshot { generation } => {
                write!(f, "{INDEX_SNAPSHOT_PREFIX}{generation}{extension}")
            }
            Self::KeyringReplica {
                generation,
                set_digest,
                replica,
            } => write!(
                f,
                "{KEYRING_PREFIX}{generation}-{set_digest}-r{}-of-{}{extension}",
                replica.index(),
                replica.count()
            ),
        }
    }
}

/// Reads a decimal number that carries no leading zeros and no sign.
fn parse_digits(digits: &str) -> Option<&str> {
    if digits.is_empty() || !digits.bytes().all(|byte| byte.is_ascii_digit()) {
        return None;
    }
    if digits.len() > 1 && digits.starts_with('0') {
        return None;
    }
    Some(digits)
}

fn parse_generation(digits: &str) -> Option<Generation> {
    parse_digits(digits)?.parse().ok().map(Generation::new)
}

fn parse_u16(digits: &str) -> Option<u16> {
    parse_digits(digits)?.parse().ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn keyring(index: u16, count: u16) -> ControlObjectName {
        ControlObjectName::keyring_replica(
            Generation::new(12),
            "a1b2",
            ReplicaPosition::new(index, count).expect("the position is valid"),
        )
        .expect("a lowercase hex digest is a valid one")
    }

    // FM-12: control objects are named `jrn-<generation>.cfrt`,
    // `idx-<generation>.cfrt`, and
    // `key-<generation>-<set_digest>-r<index>-of-<count>.cfrt`.
    #[test]
    fn names_match_the_forms_the_rule_defines() {
        assert_eq!(
            ControlObjectName::journal(Generation::new(4)).to_string(),
            "jrn-4.cfrt"
        );
        assert_eq!(
            ControlObjectName::index_snapshot(Generation::new(4)).to_string(),
            "idx-4.cfrt"
        );
        assert_eq!(keyring(1, 3).to_string(), "key-12-a1b2-r1-of-3.cfrt");
    }

    // FM-12: every form round-trips, so a name written by one device is read
    // back to the same values by another.
    #[test]
    fn every_form_round_trips() {
        let names = [
            ControlObjectName::journal(Generation::FIRST),
            ControlObjectName::journal(Generation::new(u64::MAX)),
            ControlObjectName::index_snapshot(Generation::new(9)),
            keyring(0, 1),
            keyring(2, 3),
        ];
        for name in names {
            assert_eq!(ControlObjectName::parse(&name.to_string()), Ok(name));
        }
    }

    // FM-12: Journal records and Index Snapshots use replica index 0, count 1.
    #[test]
    fn single_written_kinds_report_replica_zero_of_one() {
        for name in [
            ControlObjectName::journal(Generation::new(1)),
            ControlObjectName::index_snapshot(Generation::new(1)),
        ] {
            assert_eq!(name.replica(), ReplicaPosition::SINGLE);
            assert_eq!(name.replica().index(), 0);
            assert_eq!(name.replica().count(), 1);
            assert_eq!(name.set_digest(), None);
        }
    }

    #[test]
    fn a_name_reports_the_kind_it_belongs_to() {
        assert_eq!(
            ControlObjectName::journal(Generation::FIRST).kind(),
            ControlObjectKind::Journal
        );
        assert_eq!(
            ControlObjectName::index_snapshot(Generation::FIRST).kind(),
            ControlObjectKind::IndexSnapshot
        );
        assert_eq!(keyring(0, 1).kind(), ControlObjectKind::Keyring);
        assert_eq!(keyring(0, 1).set_digest(), Some("a1b2"));
    }

    #[test]
    fn names_outside_the_forms_are_rejected() {
        let names = [
            "jrn-4",                    // no extension
            "jrn-.cfrt",                // no generation
            "jrn-04.cfrt",              // a second spelling of generation 4
            "jrn-4x.cfrt",              // not a number
            "jrn--4.cfrt",              // signed, in effect
            "log-4.cfrt",               // not a kind coffret writes
            "key-12-a1b2-r1-of.cfrt",   // no replica count
            "key-12-a1b2-r1-to-3.cfrt", // not the `of` separator
            "key-12-a1b2-1-of-3.cfrt",  // no `r` on the index
            "key-12-zz-r1-of-3.cfrt",   // digest is not hex
            "key-12-A1B2-r1-of-3.cfrt", // digest is not lowercase
            "key-12--r1-of-3.cfrt",     // no digest
            "0011.cfrt",                // a Container, not a control object
        ];
        for name in names {
            assert_eq!(
                ControlObjectName::parse(name),
                Err(Error::MalformedObjectName {
                    name: name.to_owned()
                }),
                "{name} should not parse"
            );
        }
    }

    // FM-12: a replica index outside its count names no replica, whatever the
    // rest of the name says.
    #[test]
    fn a_name_with_an_inconsistent_replica_position_is_rejected() {
        assert_eq!(
            ControlObjectName::parse("key-12-a1b2-r3-of-3.cfrt"),
            Err(Error::Model(coffret_model::Error::InvalidReplicaPosition {
                index: 3,
                count: 3
            }))
        );
    }
}

File: frontend/packages/domain/format/src/control/objectName.ts
/**
 * The names control objects are stored under (FM-12).
 *
 * ```text
 * head-<generation>.cfrt                                 a link in the control-head chain
 * idx-<generation>.cfrt                                  an ordinary Index Snapshot
 * key-<generation>-<set_digest>-r<index>-of-<count>.cfrt  a Keyring replica
 * ```
 *
 * Control objects carry recognizable names because recovery discovers them by
 * name before any index exists.
 *
 * A name says what an object is **for**, not what it **is**: the head chain, an
 * ordinary checkpoint, a Keyring replica. Which kind an object is rides in its
 * authenticated header (FM-11), because one head position admits two kinds — the
 * ordinary Journal record and the Index Snapshot that activates a new Master Key
 * epoch both compete for the same successor slot, so naming them differently
 * would leave two keys where the commit protocol needs one (CP-2, CP-3).
 * {@link nameAdmitsKind} is the whole of that relation, and parsing a name
 * therefore yields no kind at all.
 *
 * A link in the head chain and an Index Snapshot are written once each, so their
 * names carry no replica position and they report replica index 0, count 1.
 *
 * Numbers are spelled in decimal without leading zeros, so one object has
 * exactly one name: a reader that accepted `head-007.cfrt` as generation 7 would
 * let two names claim the same object.
 */

import { U16_MAX, U64_MAX } from '../internal/bytes.js';
import { fail } from '../errors.js';
import { Generation } from '../model/generation.js';
import { ReplicaPosition } from '../model/replicaPosition.js';
import { STORAGE_EXTENSION } from '../model/containerId.js';
import type { ControlObjectKind } from '../model/kinds.js';

/** The name prefix of a link in the control-head chain. */
const HEAD_PREFIX = 'head-';
/** The name prefix of an ordinary Index Snapshot. */
const INDEX_SNAPSHOT_PREFIX = 'idx-';
/** The name prefix of a Keyring replica. */
const KEYRING_PREFIX = 'key-';

/** What an object's name says it is stored for. */
export type ControlObjectRole = 'head' | 'index-snapshot' | 'keyring-replica';

/** The name a control object is stored under. */
export interface ControlObjectName {
  /** What this name says the object is stored for. */
  role: ControlObjectRole;
  /** The generation the name encodes. */
  generation: Generation;
  /** The replica position the name encodes. */
  replica: ReplicaPosition;
  /**
   * The digest of the mapping the replica set carries, on a Keyring replica.
   *
   * Its contents are the Keyring's business; a name only needs it to be a
   * lowercase hex token, so that it cannot swallow the separators the rest of
   * the name is parsed on.
   */
  setDigest?: string;
}

/** The name of one generation of the control-head chain. */
export function headName(generation: Generation): ControlObjectName {
  return { role: 'head', generation, replica: ReplicaPosition.SINGLE };
}

/**
 * The name the successor of the head at `generation` is created under.
 *
 * Both successor kinds derive the same name from the same head, which is what
 * makes the conditional create that settles a commit a race between them rather
 * than two uncontested writes (CP-2, CP-3, FM-13).
 */
export function successorName(generation: Generation): ControlObjectName {
  return headName(generation.next());
}

/** The name of the ordinary Index Snapshot checkpointing one head. */
export function indexSnapshotName(generation: Generation): ControlObjectName {
  return { role: 'index-snapshot', generation, replica: ReplicaPosition.SINGLE };
}

/** The name of one replica of one generation of the Keyring. */
export function keyringReplicaName(
  generation: Generation,
  setDigest: string,
  replica: ReplicaPosition,
): ControlObjectName {
  return {
    role: 'keyring-replica',
    generation,
    replica,
    setDigest: requireSetDigest(setDigest),
  };
}

/** Which kinds each name form may carry (FM-12). */
const ADMITTED: Readonly<Record<ControlObjectRole, readonly ControlObjectKind[]>> = {
  head: ['journal', 'activation-snapshot'],
  'index-snapshot': ['index-snapshot'],
  'keyring-replica': ['keyring'],
};

/**
 * Whether an object of `kind` may be stored under `name` (FM-12).
 *
 * Every pairing outside this table is refused before decryption.
 */
export function nameAdmitsKind(name: ControlObjectName, kind: ControlObjectKind): boolean {
  return ADMITTED[name.role].includes(kind);
}

/** Spells a name as the string the object is stored under. */
export function formatControlObjectName(name: ControlObjectName): string {
  switch (name.role) {
    case 'head':
      return `${HEAD_PREFIX}${name.generation}${STORAGE_EXTENSION}`;
    case 'index-snapshot':
      return `${INDEX_SNAPSHOT_PREFIX}${name.generation}${STORAGE_EXTENSION}`;
    case 'keyring-replica': {
      const digest = requireSetDigest(name.setDigest);
      const position = `r${name.replica.index}-of-${name.replica.count}`;
      return `${KEYRING_PREFIX}${name.generation}-${digest}-${position}${STORAGE_EXTENSION}`;
    }
  }
}

/** Reads a name back into the values it encodes. */
export function parseControlObjectName(name: string): ControlObjectName {
  const malformed: () => never = () =>
    fail('malformed_object_name', `${JSON.stringify(name)} is not a control-object name`);

  if (!name.endsWith(STORAGE_EXTENSION)) {
    malformed();
  }
  const body = name.slice(0, -STORAGE_EXTENSION.length);

  if (body.startsWith(HEAD_PREFIX)) {
    return headName(parseGeneration(body.slice(HEAD_PREFIX.length), malformed));
  }
  if (body.startsWith(INDEX_SNAPSHOT_PREFIX)) {
    return indexSnapshotName(parseGeneration(body.slice(INDEX_SNAPSHOT_PREFIX.length), malformed));
  }
  if (!body.startsWith(KEYRING_PREFIX)) {
    malformed();
  }

  // The digest is hex, so it holds none of the `-` this splits on and the five
  // fields always land in the same places.
  const fields = body.slice(KEYRING_PREFIX.length).split('-');
  if (fields.length !== 5) {
    malformed();
  }
  const [generation, setDigest, index, of, count] = fields;
  if (of !== 'of' || !index.startsWith('r')) {
    malformed();
  }
  if (setDigest === '' || !isLowercaseHex(setDigest)) {
    malformed();
  }
  return keyringReplicaName(
    parseGeneration(generation, malformed),
    setDigest,
    ReplicaPosition.of(parseCount(index.slice(1), malformed), parseCount(count, malformed)),
  );
}

/** Whether two names name the same object. */
export function controlObjectNamesEqual(left: ControlObjectName, right: ControlObjectName): boolean {
  return (
    left.role === right.role &&
    left.generation.equals(right.generation) &&
    left.replica.equals(right.replica) &&
    left.setDigest === right.setDigest
  );
}

function requireSetDigest(setDigest: string | undefined): string {
  // Lowercase only, as every hex spelling in coffret is: two spellings of one
  // digest would be two names for one object.
  if (setDigest === undefined || setDigest === '' || !isLowercaseHex(setDigest)) {
    fail(
      'malformed_object_name',
      `a Keyring replica name carries a lowercase hex digest, found ${JSON.stringify(setDigest)}`,
    );
  }
  return setDigest;
}

function isLowercaseHex(text: string): boolean {
  return /^[0-9a-f]+$/.test(text);
}

/** Reads a decimal number that carries no leading zeros and no sign. */
function parseDigits(digits: string, malformed: () => never): bigint {
  if (!/^(0|[1-9][0-9]*)$/.test(digits)) {
    malformed();
  }
  return BigInt(digits);
}

function parseGeneration(digits: string, malformed: () => never): Generation {
  const value = parseDigits(digits, malformed);
  if (value > U64_MAX) {
    malformed();
  }
  return Generation.of(value);
}

function parseCount(digits: string, malformed: () => never): number {
  const value = parseDigits(digits, malformed);
  if (value > BigInt(U16_MAX)) {
    malformed();
  }
  return Number(value);
}
